service/common/src/main/java/org/apache/polaris/service/events/listeners/FileBufferPolarisPersistenceEventListener.java

```diff
@@ -0,0 +1,276 @@
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/** Event listener that buffers in memory and then dumps to persistence. */
+@ApplicationScoped
+@Identifier("persistence-file-buffer")
+public class FileBufferPolarisPersistenceEventListener extends PolarisPersistenceEventListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileBufferPolarisPersistenceEventListener.class);
+    MetaStoreManagerFactory metaStoreManagerFactory;
+    PolarisConfigurationStore polarisConfigurationStore;
+
+    // Key: str - Realm
+    // Value:
+    //    Key: int - shard number
+    //    Value: BufferShard - an object representing the directory and file where events are persisted on the filesystem
+    private final HashMap<String, HashMap<Integer, BufferShard>> buffers = new HashMap<>();
+    ConcurrentHashMap<String, Future> createdFlushFutures = new ConcurrentHashMap<>();
+
+    ScheduledExecutorService threadPool;
+    private static int shardCount;
+    private final int timeToFlush;
+    private final Kryo kryo = new Kryo();
+
+
+    @Inject
+    public FileBufferPolarisPersistenceEventListener(
+            MetaStoreManagerFactory metaStoreManagerFactory,
+            PolarisConfigurationStore polarisConfigurationStore
+    ) {
+        this.metaStoreManagerFactory = metaStoreManagerFactory;
+        this.polarisConfigurationStore = polarisConfigurationStore;
+        this.timeToFlush = polarisConfigurationStore.getConfiguration(null, FeatureConfiguration.EVENT_BUFFER_TIME_TO_FLUSH_IN_MS);
+        shardCount = polarisConfigurationStore.getConfiguration(null, FeatureConfiguration.EVENT_BUFFER_NUM_SHARDS);
+        int numThreads = polarisConfigurationStore.getConfiguration(null, FeatureConfiguration.EVENT_BUFFER_NUM_THREADS);
+        threadPool = Executors.newScheduledThreadPool(numThreads);
+        kryo.register(PolarisEvent.class);
+        kryo.register(PolarisEvent.ResourceType.class);
+
+        // Recover and push any orphaned files
+        discoverAndPushOrphanedFiles();
+    }
+
+    @Override
+    void addToBuffer(PolarisEvent polarisEvent, CallContext callCtx) {
+        int shardNum = polarisEvent.hashCode() % shardCount;
+        String realmId = callCtx.getRealmContext().getRealmIdentifier();
+        if (!buffers.containsKey(realmId)) {
+            createBuffersForRealm(realmId);
+        }
+        BufferShard bufferShard = buffers.get(realmId).get(shardNum);
+        if (bufferShard == null) {
+            LOGGER.error("No buffer shard found for realm: #{}, shard #{}. Event dropped: {}", realmId, shardNum, polarisEvent);
+            return;
+        }
+
+        kryo.writeObject(bufferShard.output, polarisEvent);
```

Comment:
It still seems to me there's a missing check here if the buffer has grown too large.

Reply:
I'm trying to keep this as lightweight as possible - if we're looking at buffer file sizes, then we'll have to constantly read the file metadata, which I'm really trying not to do for performance concerns

Reply:
1. You can get the number of bytes in the kryo Output for free and just track it that way
2. It's also fine if the capacity is based on a number of events rather than bytes